File: scripts/invoke_claude_agent.ts
import Anthropic from "@anthropic-ai/sdk";
import * as fs from "fs";
import * as path from "path";
import { execSync } from "child_process";

const client = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

interface AgentOptions {
  action: "generate" | "execute";
  jiraIssue?: string;
  appUrl?: string;
  testCaseFile?: string;
  selectedTestCase?: string;
  outputFile: string;
}

// Test case generation via API
async function generateTestCases(
  jiraIssue: string,
  appUrl: string
): Promise<string> {
  console.log(`[Agent] Generating test cases for ${jiraIssue}...`);

  const prompt = `You are the jira-test-generator agent.

Generate comprehensive manual test cases for Jira issue: ${jiraIssue}

Application: ${appUrl}
Test User: neuralqaacademy@gmail.com / Tester@123

Fetch the Jira issue details and create 5-10 detailed test cases in markdown format.

Each test case should include:
- Title (TC-XXX)
- Objective
- Preconditions
- Numbered test steps (plain language)
- Expected results
- Test data

Format as professional markdown documentation ONLY.
Start generating now.`;

  const response = await client.messages.create({
    model: "claude-sonnet-5",
    max_tokens: 4096,
    messages: [
      {
        role: "user",
        content: prompt,
      },
    ],
  });

  const textBlock = response.content.find((block) => block.type === "text");
  if (!textBlock || textBlock.type !== "text") {
    throw new Error("No text content in response");
  }

  return textBlock.text;
}

// Test execution via Claude CLI (has browser tools)
async function executeTestCases(
  testCaseFile: string,
  appUrl: string,
  selectedTestCase?: string
): Promise<string> {
  console.log(
    `[Agent] Executing test case: ${selectedTestCase} via Claude CLI...`
  );

  // Read test case file
  const fullContent = fs.readFileSync(testCaseFile, "utf-8");

  // Extract only the selected test case if specified
  let testCaseContent = fullContent;
  if (selectedTestCase) {
    // Extract the specific test case (e.g., TC-001)
    const testCasePattern = new RegExp(
      `## ${selectedTestCase}:.*?(?=## TC-\\d+:|$)`,
      "is"
    );
    const match = fullContent.match(testCasePattern);
    if (match) {
      testCaseContent = match[0];
      console.log(`[Agent] Extracted test case: ${selectedTestCase}`);
    } else {
      console.warn(
        `[Agent] Could not extract ${selectedTestCase}, using full file`
      );
    }
  }

  // Create a temporary prompt file
  const promptFile = path.join(
    process.env.TEMP || "/tmp",
    `execute_test_${Date.now()}.txt`
  );

  const prompt = `Execute ONLY the following test case and document results.

Application: ${appUrl}
Test User: neuralqaacademy@gmail.com / Tester@123

Test Case to Execute:
\`\`\`
${testCaseContent}
\`\`\`

Instructions:
1. Execute ONLY this ONE test case - do not execute other tests
2. Follow each test step exactly as described
3. Navigate the application and verify expected results
4. Document: Pass or Fail for each step
5. Take screenshots only if test fails
6. Document actual vs expected results

Output results in markdown format with:
- Test case ID and title
- Status: PASS or FAIL
- Step-by-step results
- Observations
- Any failures with evidence

Start executing now.`;

  fs.writeFileSync(promptFile, prompt, "utf-8");

  try {
    // Invoke Claude CLI with manual-test-runner agent (has browser tools)
    console.log("Invoking manual-test-runner agent via CLI...");
    const result = execSync(
      `claude --agent manual-test-runner "${prompt}"`,
      {
        encoding: "utf-8",
        maxBuffer: 10 * 1024 * 1024, // 10MB buffer
      }
    );

    return result;
  } catch (error) {
    console.error(
      "CLI execution error:",
      error instanceof Error ? error.message : String(error)
    );
    throw error;
  } finally {
    // Clean up temp file
    try {
      fs.unlinkSync(promptFile);
    } catch {}
  }
}

// Save content to file
async function saveContent(content: string, outputFile: string): Promise<void> {
  const outputDir = path.dirname(outputFile);

  // Create directory if it doesn't exist
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  fs.writeFileSync(outputFile, content, "utf-8");
  console.log(`✅ Saved: ${outputFile}`);
}

// Parse command line arguments
function parseArgs(): AgentOptions {
  const args = process.argv.slice(2);
  const options: Partial<AgentOptions> = { outputFile: "" };

  for (let i = 0; i < args.length; i += 2) {
    const key = args[i].replace(/^--/, "");
    const value = args[i + 1];

    switch (key) {
      case "action":
        options.action = value as "generate" | "execute";
        break;
      case "jira-issue":
        options.jiraIssue = value;
        break;
      case "app-url":
        options.appUrl = value;
        break;
      case "test-case-file":
        options.testCaseFile = value;
        break;
      case "selected-test-case":
        options.selectedTestCase = value;
        break;
      case "output-file":
        options.outputFile = value;
        break;
    }
  }

  return options as AgentOptions;
}

// Main execution
async function main(): Promise<void> {
  try {
    const options = parseArgs();

    // Validate API key
    if (!process.env.ANTHROPIC_API_KEY) {
      console.error("ERROR: ANTHROPIC_API_KEY environment variable not set");
      process.exit(1);
    }

    // Validate required options
    if (!options.action || !options.outputFile) {
      console.error(
        "ERROR: --action and --output-file are required parameters"
      );
      process.exit(1);
    }

    let content: string;

    if (options.action === "generate") {
      if (!options.jiraIssue || !options.appUrl) {
        console.error(
          "ERROR: --jira-issue and --app-url required for generate action"
        );
        process.exit(1);
      }

      content = await generateTestCases(options.jiraIssue, options.appUrl);
      await saveContent(content, options.outputFile);
      console.log("✅ Test case generation complete");
    } else if (options.action === "execute") {
      if (!options.testCaseFile || !options.appUrl) {
        console.error(
          "ERROR: --test-case-file and --app-url required for execute action"
        );
        process.exit(1);
      }

      content = await executeTestCases(
        options.testCaseFile,
        options.appUrl,
        options.selectedTestCase
      );
      await saveContent(content, options.outputFile);
      console.log("✅ Test execution complete");
    } else {
      console.error("ERROR: Invalid action. Use 'generate' or 'execute'");
      process.exit(1);
    }
  } catch (error) {
    console.error(
      "ERROR:",
      error instanceof Error ? error.message : String(error)
    );
    process.exit(1);
  }
}

main();

File: scripts/invoke_claude_agent.py
#!/usr/bin/env python3
"""
Invoke Claude agents via API for test generation and execution
Used by Jenkins to dynamically call agents
"""

import os
import sys
import json
import argparse
from anthropic import Anthropic

def invoke_test_generator(jira_issue, app_url):
    """Invoke jira-test-generator agent"""
    client = Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))

    print(f"[Agent] Generating test cases for {jira_issue}...")

    prompt = f"""You are the jira-test-generator agent.

Generate comprehensive manual test cases for Jira issue: {jira_issue}

Application: {app_url}
Test User: neuralqaacademy@gmail.com / Tester@123

Fetch the Jira issue details and create 5-10 detailed test cases in markdown format.

Each test case should include:
- Title (TC-XXX)
- Objective
- Preconditions
- Numbered test steps (plain language)
- Expected results
- Test data

Format as professional markdown.

Start generating now."""

    response = client.messages.create(
        model="claude-opus-4-1",
        max_tokens=4096,
        messages=[
            {
                "role": "user",
                "content": prompt
            }
        ]
    )

    return response.content[0].text

def invoke_test_executor(test_case_file, app_url):
    """Invoke manual-test-runner agent"""
    client = Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))

    print(f"[Agent] Executing test cases from {test_case_file}...")

    prompt = f"""You are the manual-test-runner agent.

Execute the test cases and document results.

Application: {app_url}
Test Cases: {test_case_file}
Test User: neuralqaacademy@gmail.com / Tester@123

Follow each test step:
1. Navigate the application
2. Execute each step exactly as described
3. Document: Pass or Fail for each step
4. Take screenshots only if test fails
5. Document actual vs expected results

Output results in markdown format with:
- Test case ID and title
- Status: PASS or FAIL
- Observations
- Any failures with evidence

Start executing now."""

    response = client.messages.create(
        model="claude-sonnet-4-20250514",
        max_tokens=4096,
        messages=[
            {
                "role": "user",
                "content": prompt
            }
        ]
    )

    return response.content[0].text

def save_test_cases(content, output_file):
    """Save generated test cases to file"""
    os.makedirs(os.path.dirname(output_file), exist_ok=True)

    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(content)

    print(f"✅ Saved: {output_file}")

def main():
    parser = argparse.ArgumentParser(description='Invoke Claude agents from Jenkins')
    parser.add_argument('--action', choices=['generate', 'execute'], required=True,
                        help='Action to perform')
    parser.add_argument('--jira-issue', help='Jira issue key (e.g., SCRUM-2)')
    parser.add_argument('--app-url', help='Application URL')
    parser.add_argument('--test-case-file', help='Path to test case file')
    parser.add_argument('--output-file', help='Output file path')

    args = parser.parse_args()

    # Verify API key
    if not os.environ.get("ANTHROPIC_API_KEY"):
        print("ERROR: ANTHROPIC_API_KEY environment variable not set")
        sys.exit(1)

    try:
        if args.action == 'generate':
            if not args.jira_issue or not args.app_url or not args.output_file:
                print("ERROR: --jira-issue, --app-url, and --output-file required for generate")
                sys.exit(1)

            content = invoke_test_generator(args.jira_issue, args.app_url)
            save_test_cases(content, args.output_file)
            print("✅ Test case generation complete")

        elif args.action == 'execute':
            if not args.test_case_file or not args.app_url or not args.output_file:
                print("ERROR: --test-case-file, --app-url, and --output-file required for execute")
                sys.exit(1)

            content = invoke_test_executor(args.test_case_file, args.app_url)
            save_test_cases(content, args.output_file)
            print("✅ Test execution complete")

    except Exception as e:
        print(f"ERROR: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()
